// react-native-bootpay/lib/commonjs/BootpayWebView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BootpayWebView = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _WebView = _interopRequireDefault(require("./WebView"));

var _UserInfo = _interopRequireDefault(require("./UserInfo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class BootpayWebView extends _react.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      visibility: false,
      script: '',
      firstLoad: false
    });

    _defineProperty(this, "request", (payload, items, user, extra) => {
      payload.application_id = _reactNative.Platform.OS == 'ios' ? this.props.ios_application_id : this.props.android_application_id;
      payload.items = items;
      payload.user = user;
      payload.extra = extra; //visibility가 true가 되면 webview onLoaded가 실행됨        

      this.setState({
        visibility: true,
        script: 'if(BootPay == undefined || BootPay.request == undefined) {return;} BootPay.request(' + JSON.stringify(payload) + ')',
        firstLoad: false
      });

      _UserInfo.default.updateInfo();
    });

    _defineProperty(this, "dismiss", () => {
      this.setState(({
        visibility
      }) => ({
        visibility: false
      }));
      this.removePaymentWindow();
    });

    _defineProperty(this, "onLoadEnd", async e => {
      if (this.state.firstLoad == true) return;
      this.setBootpayPlatform();
      await this.setAnalyticsData();
      this.goBootpayRequest();
      this.setState({ ...this,
        firstLoad: true
      });
    });

    _defineProperty(this, "generateScript", script => {
      const onError = '.error(function(data){ window.ReactNativeWebView.postMessage( JSON.stringify(data) ); })';
      const onCancel = '.cancel(function(data){ window.ReactNativeWebView.postMessage( JSON.stringify(data) ); })';
      const onReady = '.ready(function(data){ window.ReactNativeWebView.postMessage( JSON.stringify(data) ); })';
      const onConfirm = '.confirm(function(data){ window.ReactNativeWebView.postMessage( JSON.stringify(data) ); })';
      const onClose = '.close(function(data){ window.ReactNativeWebView.postMessage("close"); })';
      return script + onError + onCancel + onReady + onConfirm + onClose + '; void(0);';
    });

    _defineProperty(this, "onMessage", ({
      nativeEvent
    }) => {
      if (nativeEvent == undefined || nativeEvent.data == undefined) return;

      if (nativeEvent.data == 'close') {
        if (this.props.onClose == undefined) return;
        json = {
          action: 'BootpayClose',
          message: '결제창이 닫혔습니다'
        };
        this.props.onClose(json);
        this.dismiss();
        return;
      }

      const data = JSON.parse(nativeEvent.data);

      switch (data.action) {
        case 'BootpayCancel':
          if (this.props.onCancel != undefined) this.props.onCancel(data);
          break;

        case 'BootpayError':
          if (this.props.onError != undefined) this.props.onError(data);
          break;

        case 'BootpayBankReady':
          if (this.props.onReady != undefined) this.props.onReady(data);
          break;

        case 'BootpayConfirm':
          if (this.props.onConfirm != undefined) this.props.onConfirm(data);
          break;

        case 'BootpayDone':
          if (this.props.onDone != undefined) this.props.onDone(data);
          break;
      }
    });

    _defineProperty(this, "onShouldStartLoadWithRequest", url => {
      return true; // goBootpayRequest();
    });

    _defineProperty(this, "setBootpayPlatform", () => {
      if (_reactNative.Platform.OS == 'ios') {
        this.injectJavaScript("\n  BootPay.setDevice('IOS');\n          ");
      } else if (_reactNative.Platform.OS == 'android') {
        this.injectJavaScript("\n  BootPay.setDevice('ANDROID');\n          ");
      }
    });

    _defineProperty(this, "goBootpayRequest", () => {
      const fullScript = this.generateScript(this.state.script);
      this.injectJavaScript(fullScript);
    });

    _defineProperty(this, "transactionConfirm", data => {
      this.injectJavaScript("\n        var data = JSON.parse(".concat(data, "); \n        BootPay.transactionConfirm(data);\n          "));
    });

    _defineProperty(this, "removePaymentWindow", () => {
      this.injectJavaScript("\n        BootPay.removePaymentWindow();\n          ");
    });

    _defineProperty(this, "injectJavaScript", script => {
      if (this.wv == null || this.wv == undefined) return;
      this.wv.injectJavaScript("\n        javascript:(function(){".concat(script, " })()\n          "));
    });

    _defineProperty(this, "setAnalyticsData", async () => {
      const uuid = await _UserInfo.default.getBootpayUUID();
      const bootpaySK = await _UserInfo.default.getBootpaySK();
      const bootLastTime = await _UserInfo.default.getBootpayLastTime();
      const elaspedTime = Date.now() - bootLastTime;
      this.injectJavaScript(" \n        window.BootPay.setAnalyticsData({uuid:'".concat(uuid, "',sk:'").concat(bootpaySK, "',sk_time:").concat(bootLastTime, ",time:").concat(elaspedTime, "});\n        "));
    });
  }

  async componentWillUnmount() {
    this.setState({
      visibility: false,
      script: '',
      firstLoad: false
    });

    _UserInfo.default.setBootpayLastTime(Date.now());
  }

  render() {
    const injectedJavascript = "(function() {\n                window.postMessage = function(data) {\n            window.ReactNativeWebView.postMessage(data);\n            };\n        })()";
    return /*#__PURE__*/_react.default.createElement(_reactNative.Modal, {
      animationType: 'slide',
      transparent: false,
      visible: this.state.visibility
    }, /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
      style: {
        flex: 1
      }
    }, /*#__PURE__*/_react.default.createElement(_WebView.default, {
      ref: wv => this.wv = wv,
      useWebKit: true,
      originWhitelist: ['*'],
      source: {
        uri: 'https://inapp.bootpay.co.kr/3.3.1/production.html'
      },
      javaScriptEnabled: true,
      javaScriptCanOpenWindowsAutomatically: true,
      scalesPageToFit: true,
      onLoadEnd: this.onLoadEnd,
      onMessage: this.onMessage,
      onShouldStartLoadWithRequest: this.onShouldStartLoadWithRequest
    })));
  }

} // BootpayWebView.prototype = 


exports.BootpayWebView = BootpayWebView;
